### PR TITLE
HPCC-26021 Add component instrumentation section

### PR DESCRIPTION
**Pull request by kenrowland**
Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: devdoc/Metrics.rst [hunk 1]
```diff
@@ -199,22 +199,26 @@ arise in the implementation of a sink shall be the sole responsibility of the si
...
 common interface to allow the framework to manage them in a common way.
 
 To meet the requirement to manage metrics independent of the underlying metric state, all metrics
 implement a common interface. All metrics then add their specific methods to update and retrieve
-internal state. Generally the component uses the update method(s) to change metric state whenever
-an event or other process dictates. The sink, described later, is generally the consumer of the
-retrieval methods. Components create and update metrics and sinks retrieve and consume the values.
-The metric is responsible for synchronizing access between update and retrieval.
+internal state. Generally the component uses the update method(s) to update state and the framework
+uses retrieval methods to get current state when reporting. The metric insures shnchronized access.
```
Comment (ghalliday) [inline devdoc/Metrics.rst:216]: typo 'shnchronized'

File: devdoc/Metrics.rst [hunk 2]
```diff
@@ -276,6 +280,15 @@ Once created, the component shall update the gauge anytime the state of what is
 The metric shall provide methods to increase and decrease the value. The sink reads the value during
 collection and reporting.
 
+Custom Metric
+-------------
+A custom metric is a class that allows a component to leverage existing metrics. The component creates
+an instance of a custom metric (a templated class) and passes a reference to the underlying metric
+value. When collection is performed, the custom metric simply reads the value of the metric using the
+reference provided during construction. The component maintains full responsibility for updating the
+metric value as the custom metric clsas provides no update methods. The component is also responsible
```
Comment (ghalliday) [inline devdoc/Metrics.rst:289]: typo: 'clsas'

File: devdoc/Metrics.rst [hunk 3]
```diff
@@ -306,29 +316,92 @@ metrics
 metrics.sinks
     List of sinks defined for the component (may have been combined with global config)
 
-metrics.sinks.sink
-    The repeating element
-
 metrics.sinks.sink.type
```
Comment (ghalliday) [inline devdoc/Metrics.rst:319]: remove ".sink", possibly metrics.sinks[].type for clarity.  Same for following definitions.

File: devdoc/Metrics.rst [hunk 4]
```diff
@@ -306,29 +316,92 @@ metrics
...
+In order to instrument a component for metrics using the framework, a component must include the metrics
+header from jlib (*jmetrics.hpp*) and add jlib as a dependent lib (if not already doing so).
+
+The general steps for instrumentation are
+
+1. Create a metrics reporter object
+2. Create metric objects for each internal state to measure and add each to the reporter
+3. Add updates to each metric throughout the component wherever metric state changes
+
+The *metrics reporter* is a singleton created using the platform defined singleton pattern template. The compoent
```
Comment (ghalliday) [inline devdoc/Metrics.rst:345]: typo: compoent

File: devdoc/Metrics.rst [hunk 5]
```diff
@@ -306,29 +316,92 @@ metrics
...
+is assumed for all code examples that follow.
+
+::
+
+    std::shared_ptr<CounterMetric> pCounter = std::make_shared<CounterMetric>("metricName", "description");
+    metricsReporter.add(pCounter);
+
+Note the metric type for both the shared pointer variable and in the *make_shared* template that creates the
+metric and returns a shared pointer. Simply substitute other metric types and handle any differences in the
+constructor arguments as needed. It should also be noted that all metrics inherit from the *ImMtric* interface should
```
Comment (ghalliday) [inline devdoc/Metrics.rst:366]: typo: "ImMtric"

Comment (ghalliday) [reply devdoc/Metrics.rst:366]: I'm not sure this sentence is worth including.

**Review by ghalliday — CHANGES_REQUESTED** [review]
Some minor comments

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
With the exception of the typos @ghalliday already noted, this looks good.

**Comment by kenrowland** [comment]
@ghalliday I went ahead and rebased since the changes were minimal. Please take a look and let me know if there is anything else.

**Comment by ghalliday** [comment]
I re-targeted to 8.2.x.  The PR will need rebasing into the correct base branch

**Review by ghalliday — APPROVED** [review]
ok to merge once it is rebased.

**Comment by kenrowland** [comment]
Resubmitting new PR as unable to rebase to candidate-8.2.x for some reason. Unknown merge conflict un related to any local changes.

**Pull request closed without merging**